ebu_tt_live/node/denester.py

```diff
@@ -1,28 +1,71 @@
 from .base import AbstractCombinedNode
 from ebu_tt_live.documents.ebutt3 import EBUTT3Document
-from pyxb.binding.basis import NonElementContent, ElementContent
-from ebu_tt_live.bindings import div_type, p_type, span_type, ebuttdt, style_type
+from ebu_tt_live.bindings import div_type, span_type, ebuttdt, style_type
 from ebu_tt_live.bindings._ebuttm import divMetadata_type
-import copy
-import re
+from ebu_tt_live.errors import UnexpectedSequenceIdentifierError
+import logging
 from datetime import timedelta
 
+log = logging.getLogger(__name__)
+
+
 class DenesterNode(AbstractCombinedNode):
     """
-    This class contains methods for unnesting elements of an EBUTT3Document that cannot be nested in EBUTTD
+    Class to denest div and span elements.
+
+    Whereas EBU-TT Part 1 and EBU-TT Part 3 permit div elements
+    as children of div elements, and span elements of children of
+    span elements, such nested elements are not permitted in other
+    profiles, such as EBU-TT-D. This class provides methods for
+    flattening such nested elements, for example for use prior to
+    conversion to EBU-TT-D.
     """
 
+    _sequence_identifier = None
     _expects = EBUTT3Document
     _provides = EBUTT3Document
 
-    def __init__(self, node_id, consumer_carriage=None, producer_carriage=None):
+    def __init__(self,
+                 node_id,
+                 sequence_identifier,
+                 consumer_carriage=None,
+                 producer_carriage=None):
         super(DenesterNode, self).__init__(
             node_id=node_id,
             producer_carriage=producer_carriage,
             consumer_carriage=consumer_carriage
         )
+        self._sequence_identifier = sequence_identifier
 
     def process_document(self, document, **kwargs):
+        if self.is_document(document):
+
+            if document.sequence_identifier == self._sequence_identifier:
+                raise UnexpectedSequenceIdentifierError()
+
+            if self.check_if_document_seen(document=document):
+
+                self.limit_sequence_to_one(document)
+
+                # Change the sequence identifier
+                document.sequence_identifier = self._sequence_identifier
+
+                # Do the denesting
+                document = DenesterNode.denest(document)
+
+                document.validate()
+                self.producer_carriage.emit_data(data=document, **kwargs)
+            else:
+                log.warning(
+                    'Ignoring duplicate document: {}__{}'.format(
+                        document.sequence_identifier,
+                        document.sequence_number
+                    )
+                )
+        else:
+            self.producer_carriage.emit_data(data=document, **kwargs)
+
+    def denest(document):
```

Comment:
Shouldn't all these functions have `self` as their first argument? Alternatively you should be able to move the function declarations out of the class definition.

Reply:
It seems like a philosophical question. The functions are gathered into a class for convenience and addressing, but none of them needs access to `self` so why make them have it? I'd be more strongly in favour of a change if there were some data that needed to be stored in the object and accessed in the functions, but I couldn't see any!

Reply:
By the way, they don't even need `cls`, either.

Reply:
Looking at another e.g. `delay.py`, it has functions declared outside the class. They'll still be in the same python file so it's still easy to reference.

It just seems odd having all these warnings pop up on my screen, makes it seem like this sort of pattern isn't standard practice even though it seems to work.

Reply:
Discussion result: add `@staticmethod` decorator to all the static methods.

Reply:
Fixed in f465381